### ISSUE TITLE
SOFA delay before Nudges start

### ISSUE THREAD

**Issue by kevinmcox** [macadmins/nudge#573]
Requesting a  feature related to the new SOFA functionality that provides a delay before Nudge starts Nudging.

e.g. New release detected; wait 72 hours to start Nudging; enforce SLA of 14 days (from release date).

This would allow admins to take advantage of the new automatic SOFA functionality but also have a configurable delay in case issues are detected by early adopters.

**Comment by erikng** [comment]
So just to confirm, you want the SLA to be 14 days from the release date, but with 3 days grace. So is that 17 days from release or just 11 days left when nudge finally activates?

**Comment by BigMacAdmin** [comment]
I would vote that the SLA value becomes "the duration of the campaign" it feels more natural, but probably makes the math a bit more difficult on your side.

Which way would result in fewer confused admins asking about it on Slack? :)

**Comment by kevinmcox** [comment]
I can adapt to however you want to implement it (or think is the most logical for everyone).

The way we use Nudge right now is that we wait 72 hours after a macOS release to push the new Nudge config profile. This profile has a deadline 7 days in the future.

So effectively we enforce a 10 day deadline on new macOS releases, but only have an "active Nudge cycle" for the last 7 of those 10 days.

So in my mind, the SLA is based on the days since release.

The delay/mute does NOT extend the deadline, just keeps Nudge quiet until that threshold is passed.

This feels more straight forward to configure and doesn't require doing math to calculate your deadline if you change the delay.

**Comment by erikng** [comment]
Sounds like we are in agreement even though I hadn't said my position :)

I think if we are going to implement a delay it needs to be mindful of the SLAs to not overly punish the users. We've done that in the past (not deploy a new nudge) and users do feel like they got the short end of the stick.

**Comment by erikng** [comment]
By default this key will be 0 days

**Comment by kevinmcox** [comment]
> I think if we are going to implement a delay it needs to be mindful of the SLAs to not overly punish the users. We've done that in the past (not deploy a new nudge) and users do feel like they got the short end of the stick.

Do you want to try to account for this in code? Or just document the concern and put the burden on admins to not hang themselves?

**Comment by erikng** [comment]
I can handle it. Should be easy to implement and give admin flexibility.

**Comment by Tribruin** [comment]
I would suggest that Nudge also check the preferences settings for enforcedSoftwareUpdateDelay, enforcedSoftwareUpdateMajorOSDeferredInstallDelay, and enforcedSoftwareUpdateMinorOSDeferredInstallDelay in the com.apple.applicationaccess preference domain and use those to calculate whether to display the Nudge prompt.

We have different settings for different users, so some users may only be deferred for 2 days and other users for 5 days. Having Nudge account for the built-in delay would be helpful, otherwise, we would need separate profile settings.

**Comment by erikng** [comment]
Sorry I don't plan on doing that. That's going to get incredibly complex and add more dependencies to systems that I then have to account for when they aren't there.

**Comment by erikng** [comment]
The next pre-release of nudge will have this `sofaPeriodLaunchDelay`.

**Comment by kevinmcox** [comment]
@erikng I don't see `sofaPeriodLaunchDelay` on the CHANGELOG:
https://github.com/macadmins/nudge/blob/development/CHANGELOG.md

Did this key get renamed?

The description of `nudgeEventLaunchDelay` doesn't seem to match what we talked about here.

**Comment by erikng** [comment]
I did in fact rename it to support another request. This is the exact same code but used to compute all delays on the required installation date.

**Comment by erikng** [comment]
The SLA logic is separate and computed against this property which is how the code still works even though the name has changed.

**Comment by kevinmcox** [comment]
Gotcha. So an SLA of `10` days, combined with `nudgeEventLaunchDelay` of `3` days, will result in Nudges starting three days after release with the deadline 10 days after release. Is that correct?

**Comment by erikng** [comment]
Yep. Please test and confirm my brain wrote it all correctly though.

**Comment by kevinmcox** [comment]
This seems to work as expected. One note is that Nudge does not appear to log any note when the interface is not displayed due to `nudgeEventLaunchDelay`.

**Comment by erikng** [comment]
It does log but as an info event. Many of those events are not capture by the daemon to reduce its size but you should see it in console.app

**Comment by kevinmcox** [comment]
I don't see it there either, but maybe I'm just missing it?
